NED Refactoring (4)

Mehrad0711 opened the pull request.

Mehrad0711 committed:
Update oracle

Mehrad0711 committed:
Decouple db and task

db data processing (similar to bootleg) is task independent and should run outside of task

Mehrad0711 committed:
Database fathers Bootleg

Bootleg does entity disambiguations and Database does entity linking. Both can be subclasses of a more general class (NED) but for now let Database father Bootleg.

Mehrad0711 committed:
Add db processing to server and predict

Mehrad0711 committed:
Remove separate branching in run_bootleg

Mehrad0711 committed:
Fix bug for embedding based NED

Use padding features for non-utterance field. Both will be concatentaed later.

Mehrad0711 committed:
Rename canonical2type.json to alias2type.json

Mehrad0711 committed:
Drop almond_thingtalk_version argument

it was used in oracle experiments to extract entities and property names based on syntax.
as most previous almond datasets have moved to thinktalk 2.0 we can drop support for 1.0

Mehrad0711 committed:
Define AbstractEntityLinker

All classes that process features for examples extend this class. It contains main methods to process entity features and add them to Example object.
Other classes would write their own process_examples method.

Mehrad0711 committed:
Reorganize NED code

Move all ned and nel models to ned directory. Refactor the code so we instantiate an ned_model based on ned_retrieve_method argument.

Mehrad0711 committed:
bootleg: add dump_entities_with_labels method

which run_bootleg.py would call to dump entities and their labels

Mehrad0711 committed:
Relative import all modules

gcampax commented on README.md:
````diff
@@ -149,7 +149,7 @@ It will also generate several files in <results_temp> folder. In `eval_bootleg/[
 Now you can use the extracted features from bootleg in downstream tasks such as semantic parsing to improve named entity understanding and consequently generation:
 ```bash
-genienlp train --train_tasks <train_task_names> --train_iterations <iterations> --preserve_case --save <savedir> --data <dataset_dir> --model TransformerSeq2Seq --pretrained_model facebook/bart-base --train_batch_tokens 1000 --val_batch_size 1000 --do_ned --database_dir <database_dir> --ned_retrieve_method bootleg --entity_attributes type_id type_prob --add_entities_to_text append --bootleg_model ...
+genienlp train --train_tasks <train_task_names> --train_iterations <iterations> --preserve_case --save <savedir> --data <dataset_dir> --model TransformerSeq2Seq --pretrained_model facebook/bart-base --train_batch_tokens 1000 --val_batch_size 1000 --do_ned --database_dir <database_dir> --ned_class_name bootleg --entity_attributes type_id type_prob --add_entities_to_text append --bootleg_model <boo ...
````
Please undo this change, ned_retrieve_method is a much more informative argument.

gcampax commented on genienlp/ned/ned_utils.py:
```diff
@@ -281,18 +281,18 @@ def init_ned_model(args, ned_class_name=None):
+            if args.ned_class_name == 'bootleg':
                 ned_class_name = 'Bootleg'
-            elif args.ned_retrieve_method == 'bootleg-annotator':
+            elif args.ned_class_name == 'bootleg-annotator':
```
ned_retrieve_method can never be bootleg-annotator, right? The difference between bootleg and bootleg-annotator is whether you're called from "server" or "train" / "predict"

Mehrad0711 replied:
yes. that's why there's the option of passing `ned_class_name` as an argument to override what args say.
Alternatively, we can pass a `mode` which can be (server|batch) but I avoided that since server mode and batch mode for non bootleg classes is the same.

Mehrad0711 committed:
Several renamings

Bootleg --> BatchBootlegEntityDisambiguator
BootlegAnnotator --> ServingBootlegEntityDisambiguator
XXXEntityLinker --> XXXEntityDisambiguator

Mehrad0711 committed:
ned: minor fixes

gcampax commented on genienlp/ned/base.py:
```diff
@@ -1,5 +1,5 @@
 #
-# Copyright (c) 2020-2021 The Board of Trustees of the Leland Stanford Junior University
+# Copyright (c) 2021-2022 The Board of Trustees of the Leland Stanford Junior University
```
2022?

gcampax commented on genienlp/ned/base.py:
```diff
@@ -0,0 +1,333 @@
+            for title in titles:
+                self.wiki2normalized_type.append((title, normalized_type))
+
+    def process_examples(self, examples, split_path, utterance_field):
```
How can this method have an implementation if this is an abstract class?

Mehrad0711 replied:
moved to base class

gcampax commented on genienlp/ned/base.py:
```diff
@@ -0,0 +1,333 @@
+
+        return tokens_type_ids
+
+    def lookup(self, tokens, database_lookup_method=None, min_entity_len=2, max_entity_len=4):
```
This method is not used by Bootleg right? So it doesn't belong here. It should go into a subclass.

Mehrad0711 replied:
Ok yeah I should create a base class for Naive and Oracle and many of these methods should move there.

gcampax commented on genienlp/ned/bootleg.py:
```diff
@@ -30,80 +30,44 @@
-        self.model_ckpt_path = os.path.join(self.model_dir, self.ckpt_name + '.' + extension)
+        ###
+
+        self.ckpt_name = 'bootleg_wiki'
```
This is a cmdline argument right?

Mehrad0711 replied:
Checkpoint name used to be different for different bootleg models. Now it's all the same. We still can pass different bootleg_models.

gcampax commented on genienlp/ned/bootleg.py:
```diff
@@ -173,21 +132,16 @@ def extract_mentions(self, input_path):
     def disambiguate_mentions(self, config_args):
         run_model(self.args.bootleg_dump_mode, config_args)
 
     def post_process_bootleg_types(self, title):
```
Why is the initialization for type postprocessing done in the parent class, but type postprocessing is only used in Bootleg?

Mehrad0711 replied:
Oracle uses `wiki2normalized_type` to normalize types too. Except only uses the first half of this function.  I have defined a base `normalize_types` method that other classes would inherit in a new PR I'm working on. That would solve this confusion.

gcampax commented on genienlp/ned/bootleg.py:
```diff
@@ -323,103 +215,70 @@ def collect_features_per_line(self, line, threshold):
+        self.disambiguate_mentions(config_args)
+
+
+class ServingBootlegEntityDisambiguator(BatchBootlegEntityDisambiguator):
```
Do you need this inheritance? What are you inheriting?

Mehrad0711 replied:
We discussed this on slack. `ServingBootlegEntityDisambiguator` uses several methods and attributes from `BatchBootlegEntityDisambiguator`. Turning them into utility functions is not viable since methods are called within each other.

gcampax commented on genienlp/ned/main.py:
```diff
@@ -0,0 +1,190 @@
+logger = logging.getLogger(__name__)
+
+
+class NaiveEntityDisambiguator(AbstractEntityDisambiguator):
```
How can the classes in this file not implement process_examples?

gcampax commented on genienlp/ned/main.py:
```diff
@@ -0,0 +1,190 @@
+        super().__init__(args)
+
+        with open(os.path.join(self.args.database_dir, 'es_material/alias2type.json'), 'r') as fin:
+            # alias2type.json is a big file (>4G); load it only when necessary
```
Why is alias2type a JSON file at all, and not a binary-serialized marisa trie?

Mehrad0711 replied:
Good point. I will convert it.

gcampax reviewed (changes requested):
This is definitely a good cleanup, and we're getting there, but we can still do better.

Mehrad0711 committed:
Address PR comments

Mehrad0711 committed:
bootleg: fix sorting bug
recent changes in bootleg repository removed sorting output labels based on candidate probability
this backward incompatible change caused silent bugs here. Do a temporary fix by sorting labels in side genienlp which ultimately should be added again to bootleg repo.

gcampax reviewed (approved):
Alright, I think this is a very nice cleanup, let's get it in!
